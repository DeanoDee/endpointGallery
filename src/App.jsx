import React, {Component} from 'react';
import {EndpointEntry} from './endpointEntry/';

class App extends Component {

	componentWillMount() {
		let endpoint = '';
		this.setState({endpoint});
	}

	changeEndpoint = (endpoint) => {
		console.log('endpoint changed', endpoint);
		this.setState({endpoint});
	}

	render() {
		let {changeEndpoint} = this;
		return (
			<div className="App">
				<EndpointEntry changeEndpoint={changeEndpoint}/>
				<p className="App-intro">
					To get started, edit
					<code>src/App.js</code>
					and save to reload.
				</p>
			</div>
		);
	}
}

export default App;
