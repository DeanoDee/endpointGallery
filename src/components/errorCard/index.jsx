import React from 'react';
import './index.less';

export const ErrorCard = ({error, endpoint}) => {
	return (
		<div className="error-card">
			<h2>Something went wrong</h2>
			<p>The machine said, <strong>"{error}"</strong>. I guess it didn't like <strong><em>{endpoint}</em></strong> as a place to get images.</p>
		</div>
	);
};
