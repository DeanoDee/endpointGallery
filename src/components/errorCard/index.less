@import "../../styles/main";

.error-card {
	#grid > .centeredTextCard();
	margin-top: @spacing-400;
	background: @white;
	border-radius: @border-radius;
	box-shadow: @spacing-50 @spacing-50 0 @blue-green;
	padding: @spacing-100 @spacing-200;
	> h2 {
		#typeography > .type-500;
		color: @neutral-700;
		text-align: center;
	}
	> p {
		#typeography > .type-300;
		color: @neutral-600;
	}
}
