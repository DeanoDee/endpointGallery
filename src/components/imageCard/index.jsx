import React, {PropTypes} from 'react';
import EllipsisText  from 'react-ellipsis-text';
import './index.less';

export const ImageCard = ({imageObject}) => {

	const maxCharacterLength= 78;

	const getTagElement = (tag) => {
		return (
			<li key={tag}>{tag}</li>
		);
	};

	const getTags = (tags) => {
		if(!tags) {
			return null;
		}
		return (
			<div className="tag-wrapper">
				<h3>Tags:</h3>
				<ul>{tags.map(getTagElement)}</ul>
			</div>
		);
	};

	const {title, image, description, tags} = imageObject;

	return (
		<div className="image-card">
			<div className="wrapper">
				<h2>{title}</h2>
				<div className="image-frame">
					<img src={image} alt={title}/>
				</div>
				<EllipsisText text={description} length={maxCharacterLength} />
				{getTags(tags)}
			</div>
		</div>

	);

};

ImageCard.propTypes = {
	imageObject: PropTypes.object.isRequired
};
