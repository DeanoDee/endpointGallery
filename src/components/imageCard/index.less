@import "../../styles/main";

.image-card {
	@image-height: 200px;

	#grid > .typtic();
	margin-bottom: @spacing-200;
	.wrapper {
		background: @white;
		padding: @spacing-100;
		border-radius: @border-radius;
		box-shadow: @spacing-50 @spacing-50 0 @blue-green;
		> h2 {
			#typeography > .type-400();
			font-family: @playfair;
			margin-bottom: @spacing-100;
			color: @blue-green;
			@media only screen and (min-width: @main-column-sm) {
				#typeography > .type-500();
			}
		}
		> .image-frame{
			position: relative;
			width: 100%;
			height: @image-height;
			margin-bottom: @spacing-100;
			overflow: hidden;
			background-color: @blue-light;
			> img {
				position: absolute;
				top: 0;
				transform: translate(0, -15%);
				mix-blend-mode: color-burn;
				@media only screen and (min-width: @main-column-md) {
					top: 50%;
					transform: translate(0, -50%);
				}
				width: 100%;
			}
		}
		> span {
			#typeography > .type-300();
			color: @neutral-700;
			margin-bottom: @spacing-100;
		}

		> .tag-wrapper {
			> h3 {
				color: @neutral-800;
				#typeography > .type-300();
			}

			> ul {
				list-style: none;
				padding: 0;
				margin: 0;
				> li{
					display: inline;
					color: @rose;
					&:after {
						content: ", ";
						color: @neutral-400;
					}
					&:last-child:after {
						content: "";
					}
				}

			}
		}
	}

}
