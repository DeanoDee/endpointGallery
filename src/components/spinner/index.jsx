import React from 'react';
import FontAwesome from 'react-fontawesome';
import './index.less';

export const Spinner = () => {
	return (
		<div className="spinner">
			<FontAwesome name="refresh" className="fa-spin" />
		</div>
	);
};
