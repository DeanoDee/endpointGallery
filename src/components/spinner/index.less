@import "../../styles/main";

.spinner {
	#grid > .full();
	#typeography > .type-700;
	text-align: center;
	margin-top: @spacing-400;
	color: @blue-green;
}
