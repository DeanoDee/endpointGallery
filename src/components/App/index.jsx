import React, {Component} from 'react';
import {EndpointEntry} from '../endpointEntry';
import {Logo} from '../logo';
import {ImageCard} from '../imageCard';
import {Spinner} from '../spinner';
import {ErrorCard} from '../errorCard';
import './index.less';
import {poll} from '../../lib/polling.js';


class App extends Component {

	constructor() {
		super();
		const {host} = window.location;
		const endpoint = `http://${host}/defaultImages.json`;
		this.state = {endpoint, loading: true, data: null, error: null};
	}

	componentWillMount() {
		const {getData} = this;
		getData();
	}

	changeEndpoint = (endpoint) => {
		const loading = true;
		const error = null;
		const {subscription} = this.state;
		if(subscription){
			subscription.unsubscribe();
		}
		this.setState({endpoint, error, loading}, this.getData);
	}

	setData = (object) => {
		const loading = false;
		const first = Object.keys(object)[0];
		const data = object[first];
		this.setState({data, loading});
	}

	getData = () => {
		const {setData, setError, state} = this;
		const {endpoint} = state;
		const subscription = poll(endpoint).subscribe(setData, setError);
		this.setState({subscription});
	}

	setError = (error) => {
		error = error.toString();
		const loading = false;
		this.setState({loading, error});
	}

	getCards = (imageObject) => {
		const {_id} = imageObject;
		return (
			<ImageCard key={_id} imageObject={imageObject}/>
		);
	}

	getBody = ({loading, error, data, endpoint}, getCards) => {
		if (loading) {
			return (
				<Spinner />
			);
		} else if (error){
			return (
				<ErrorCard error={error} endpoint={endpoint} />
			);
		}
		return data.map(getCards);
	}

	render() {
		const {changeEndpoint, getBody, getCards, state} = this;
		return (
			<div className="app">
				<Logo />
				<EndpointEntry changeEndpoint={changeEndpoint}/>
				{getBody(state, getCards)}
			</div>
		);
	}
}

export default App;
