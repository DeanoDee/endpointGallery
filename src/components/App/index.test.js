import React from 'react';
import ReactDOM from 'react-dom';
import App from './index';
import TestUtils from 'react-addons-test-utils';
import {Spinner} from '../spinner';
import {ErrorCard} from '../errorCard';
import {ImageCard} from '../imageCard';

describe('App', () => {
	const app = TestUtils.renderIntoDocument(<App/>);
	const mockEndpoint = "http://localhost:3002/alternateImages.json";
	const badMockEndpoint = "asdfafasf";
	const mockData = {
		"name": [
			{
				"_id": "N0tv3ryR3a0ab13Gu1d",
				"title": "A title for the image",
				"image": "http://somedomain.com/a/url/for/the/image.jpg",
				"description": "A little something to tell the person what they are looking at",
				"tags": [
					"somethings",
					"to",
					"describe",
					"what",
					"this",
					"is"
				]
			}
		]
	};
	const mockError = "SomeError";

	it('renders without crashing', () => {
		const div = document.createElement('div');
		ReactDOM.render(<App/>, div);
	});
	it('renders correctly', () => {
		const renderer = TestUtils.createRenderer();
		renderer.render(<App />);
		const tree = renderer.getRenderOutput();
		expect(tree).toMatchSnapshot();
	});
	it('changeEndpoint sets the endpoint and loading to true, and clears errors', () => {
		app.changeEndpoint(mockEndpoint);
		const {endpoint, loading, error} = app.state;
		expect(endpoint).toBe(mockEndpoint);
		expect(loading).toBe(true);
		expect(error).toBe(null);
	});
	it('setData sets the data correctly and loading to false', () => {
		app.setData(mockData);
		const {data, loading} = app.state;
		expect(data).toBe(mockData.name);
		expect(loading).toBe(false);
	});
	it('setError set the error to the passed message and loading to false', () => {

		app.setError(mockError);
		const {error, loading} = app.state;
		expect(error).toBe(mockError);
		expect(loading).toBe(false);
	});
	it('getBody with loading true displays the spinner, not error card and does not run getCards', () => {
		const getCards = jest.fn();
		const state = {
			loading: true,
			data: mockData.name,
			error: true,
			endpoint: mockEndpoint
		};
		const result = app.getBody(state, getCards);
		expect(result.type).toBe(Spinner);
		expect(result.type).not.toBe(ErrorCard);
		expect(getCards.mock.calls.length).toBe(0);
	});
	it('getBody with loading false and error has value displays the error, not spinner and does not run getCards', () => {
		const getCards = jest.fn();
		const state = {
			loading: false,
			data: mockData.name,
			error: mockError,
			endpoint: mockEndpoint
		};
		const result = app.getBody(state, getCards);
		expect(result.type).toBe(ErrorCard);
		expect(result.type).not.toBe(Spinner);
		expect(getCards.mock.calls.length).toBe(0);
	});
	it('getBody with data, loading and error false does not display spinner or error card and run getCards', () => {
		const getCards = jest.fn();
		const state = {
			loading: false,
			data: mockData.name,
			error: false,
			endpoint: mockEndpoint
		};
		const result = app.getBody(state, getCards);
		expect(result.type).not.toBe(ErrorCard);
		expect(result.type).not.toBe(Spinner);
		expect(getCards.mock.calls.length).toBe(1);
	});
	// it('getCards with returns an image card', () => {
	// 	const result = app.getCards(mockData.name);
	// 	expect(result.type).toBe(ImageCard);
	// });
});
