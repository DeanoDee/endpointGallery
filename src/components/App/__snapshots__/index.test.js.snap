exports[`App renders correctly 1`] = `
<div
  className="app">
  <Unknown />
  <EndpointEntry
    changeEndpoint={[Function]} />
  <Unknown />
</div>
`;
